src/ir_pattern.py

```diff
@@ -0,0 +1,443 @@
+        for pat_reg, pat_refs in pat.items():
+            # For now, skip mapping any memory accesses outside of the stack.
+            # Usually, these accesses are not intended to be part of the matched pattern,
+            # but in the future the pattern syntax could be extended to explicitly mark
+            # which accesses must be matched.
+            if (
+                isinstance(pat_reg, MemoryAccess)
+                and pat_reg.base_reg != self.arch.stack_pointer_reg
+            ):
+                continue
+            cand_reg = self.map_access(pat_reg)
+            cand_refs = cand.get(cand_reg)
+            if not self.match_refset(pat_refs, cand_refs):
+                return False
+        return True
+
+
+def simplify_ir_patterns(
+    arch: ArchFlowGraph, flow_graph: FlowGraph, pattern_classes: List[Type[IrPattern]]
+) -> None:
+    # Precompute a RefSet for each mnemonic
+    refs_by_mnemonic = defaultdict(list)
+    for node in flow_graph.nodes:
+        for i, instr in enumerate(node.block.instructions):
+            ref = InstrRef(node, i)
+            refs_by_mnemonic[instr.mnemonic].append(ref)
+
+    def replace_instr(ref: InstrRef, new_asm: AsmInstruction) -> None:
+        # Remove ref from all instr_references
+        instr = ref.instruction()
+        for rs in flow_graph.instr_inputs[ref].values():
+            for r in rs:
+                if isinstance(r, InstrRef):
+                    flow_graph.instr_references[r].remove_ref(ref)
+
+        # Parse the asm & set the clobbers
+        new_instr = arch.parse(new_asm.mnemonic, new_asm.args, instr.meta.derived())
+        new_instr.clobbers.extend(
+            acc for acc in instr.outputs if acc not in new_instr.outputs
+        )
+        new_instr.clobbers.extend(
+            acc for acc in instr.clobbers if acc not in new_instr.clobbers
+        )
+
+        # Replace the instruction in the block
+        ref.node.block.instructions[ref.index] = new_instr
+
+    for pattern_class in pattern_classes:
+        pattern = pattern_class.compile(arch)
+
+        # For now, patterns can't have branches: they should only have 2 Nodes,
+        # a BaseNode and an (empty) TerminalNode
+        assert (
+            len(pattern.flow_graph.nodes) == 2
+        ), "branching patterns not yet supported"
+        assert isinstance(pattern.flow_graph.nodes[0], BaseNode)
+        assert isinstance(pattern.flow_graph.nodes[1], TerminalNode)
+        pattern_node = pattern.flow_graph.nodes[0]
+
+        # Perform a brute force-ish graph search to find candidate sets of instructions
+        # that match the pattern with the correct dependencies & arguments. This will
+        # have poor performance with large patterns that use common mnemonics.
+        partial_matches = [TryIrMatch(arch=arch)]
+        for i, pat_instr in enumerate(pattern_node.block.instructions):
+            if pat_instr.mnemonic == "nop":
+                continue
```

Comment:
huh? why do nops even show up in patterns?

Reply:
I use fake `nop` instructions in `IrPattern.compile` to act as placeholder references for inputs & outputs. 
I could make this more clear by using different mnemonic name(s), but I would need to add them to `MipsArch`/`PpcArch`?